### PR TITLE
Classify errors as bugs vs runtime input errors

### PR DESCRIPTION
**Pull request by elazarg**
## Summary

- Introduces a small, deliberately shallow error hierarchy (`PrevailError`, `RuntimeInputError`, `InternalError`) in a new dependency-free header `crab_utils/prevail_errors.hpp`. The single load-bearing distinction is **bug vs runtime input error**; sub-classifying `RuntimeInputError` further is deferred until a handler differentiation justifies it.
- `main.cpp`'s top-level catch arms now distinguish the two: `InternalError` exits 2 with a "this is a bug; please file an issue" message; `RuntimeInputError` exits 1 with the existing `error: ...` shape.
- Adds missing handling at input boundaries (YAML file-open check, yaml-cpp wrapping with line/column context, assembly-parser numeric and lookup wrappers).
- Drops dead code (`SafeI64` / `num_safeint.hpp`) in a separate commit.

## Why

The codebase mixed exceptions, variants, optionals, catch-alls, and bug-style throws. A user seeing `error: ...` from `main.cpp` could not tell whether their input was bad or Prevail had a bug. `CRAB_ERROR` and `InvalidControlFlow` were both used for either category. `VerificationError` derived from `std::runtime_error` despite never being thrown (only stored in `optional<>`). The goal here is *classification only* — making the categories greppable, typed, and surfaced distinctly to the user. The eventual policy decision (abort vs throw vs `std::expected` vs contracts) is left for follow-up work.

## What changed

**Hierarchy and routing** (`crab_utils/prevail_errors.hpp`, `crab_utils/debug.hpp`)
- `PrevailError`, `RuntimeInputError`, `InternalError` defined in a new neutral header.
- `CRAB_ERROR(...)` now throws `InternalError` instead of bare `std::runtime_error`.
- `UnmarshalError` and `InvalidControlFlow` reparented under `RuntimeInputError`.
- Decorative `MalformedElf` collapsed into `UnmarshalError` (never separately caught anywhere).

**Throw-site classification**
- Generic `runtime_error` throws at user/input boundaries → `RuntimeInputError`: `printing.cpp` output-file open, `ir/parse.cpp` constraint parsing, `linux/linux_platform.cpp` map creation, `crab/type_domain.cpp` unsupported type names.
- Residual invariant-style throws → `CRAB_ERROR`: `string_constraints.hpp` "cannot iterate bottom", `printing.cpp` label lookup during disassembly, `crab/ebpf_transformer.cpp` `logic_error`, the `cfg_builder.cpp` site whose own message said "internal error" but threw `InvalidControlFlow`.
- `ir/marshal.cpp` IR-shape checks → `CRAB_ERROR`. The marshaller has no production callers — it is exercised only by `test/test_marshal.cpp` round-trip oracles, and reaching these throws means a test constructed an IR shape that doesn't have a valid bytecode encoding (a Prevail-side gap, not user input).

**`VerificationError` becomes a plain struct**
- It was a value type pretending to be an exception (`: std::runtime_error`, never actually caught polymorphically). Now: `struct VerificationError { std::string message; std::optional<Label> where; }`.
- The local non-local-exit inside `EbpfChecker` is preserved by introducing a small file-private `VerificationFailureSignal` exception, converted at the catch site into the value-typed `VerificationError`.

**Missing handling added**
- `read_suite()` checks the `ifstream` open before `YAML::LoadAll` (a missing test file used to silently produce zero documents).
- yaml-cpp parsing in `read_suite` is wrapped to preserve line/column context with the suite path.
- Assembly parser gains `parse_with_context` wrappers around `stoi`/`stoll`/`stoull` and named lookup helpers (`lookup_op`, `lookup_label`) so errors point at the offending token instead of leaking generic `"stoi"` / `"map::at"` `what()` strings.

**Other small improvements**
- `RejectionReason` consolidation: added an `enforce_instruction_feature_support` helper so the value-returning checker remains pure-query while the throw-shape conversion happens in one named place.
- `[[nodiscard]]` added to `unmarshal()`, `ebpf_domain_check`, `check_loop_bound`, `check_instruction_feature_support`.
- Dead `num_safeint.hpp` deleted (zero references; its `checked_div` divided before checking for zero).

## Out of scope

- The error-handling *policy* itself (#1079) — abort vs throw vs `std::expected` vs contracts. This PR makes the categories greppable so the policy decision can be made site-by-site without re-auditing throw sites.
- Sub-classifying `RuntimeInputError` (`IoError`, `UnsupportedFeature`, `InvalidProgramStructure`) — deferred until a handler differentiation justifies it.
- Untangling test-helper string-parsing (`parse_linear_constraints`) from the production API of `EbpfDomain` / `AnalysisResult` — tracked in #1103.
- `num_big.hpp` `Number` parse-string and `narrow<T>` family — these `CRAB_ERROR` sites are reachable from input-driven values and may belong on the runtime-input side per call site; left alone for case-by-case reclassification.
- Public-API C++ standard floor — owned by the public-API stability workstream.
- Relocation skip in `elf_reader.cpp:641-643` — plausibly intentional for multi-program reloc sections; needs a comment + targeted test before changing behavior.

## Test plan

- [x] `cmake --build build --parallel` clean.
- [x] `./bin/tests` passes at the documented baseline: `1557 cases = 1320 passed + 236 failed-as-expected + 1 skipped`, `8842 assertions = 8606 passed + 236 failed-as-expected`. Zero unexpected failures.
- [x] DCO sign-off, license headers, and project formatter all pass on the commits.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/main.cpp [hunk 1]
```diff
@@ -166,6 +166,10 @@ int main(int argc, char** argv) {
     if (!list) {
         try {
             raw_progs = elf.get_programs(desired_section, desired_program);
+        } catch (const InternalError& e) {
+            std::cerr << "internal error: " << e.what() << std::endl;
+            std::cerr << "this is a bug in Prevail; please file an issue." << std::endl;
+            return 2;
         } catch (const std::runtime_error& e) {
             load_error = e.what();
         }
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/main.cpp:178]: Fixed in bec90cce. Both blocks now catch `InternalError` first (exit 2), `RuntimeInputError` next (existing behavior), and a generic `std::exception` fallback that prints `error: ...` and exits 1 — so any non-categorized exception no longer gets quietly recorded as `load_error`.

File: src/test/ebpf_yaml.cpp [hunk 2]
```diff
@@ -360,25 +360,33 @@ static TestCase read_case(const RawTestCase& raw_case) {
...
+                TestCase tc;
+                tc.name = raw_case.test_case;
+                tc.expected_exception = raw_case.expected_exception;
+
+                try {
+                    (void)read_case(raw_case);
+                    tc.actual_exception = std::nullopt;
+                } catch (const std::exception& ex) {
+                    tc.actual_exception = ex.what();
+                }
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/test/ebpf_yaml.cpp:382]: Fixed in bec90cce. Narrowed to `catch (RuntimeInputError&)` + `catch (YAML::Exception&)` so `InternalError` is no longer eligible to satisfy an `expected-exception` clause. The schema-validation throws inside `parse_observations` and `raw_options_to_options` were retagged from `std::runtime_error` to `RuntimeInputError` so the narrowed catch still matches them (otherwise `test-data/observe.yaml` regressed).

**Comment by elazarg** [comment]
Addressed all three CodeRabbit findings in bec90cce:

1. `main.cpp:169-175` and `:196-203` — Narrowed both blocks to `catch (RuntimeInputError&)` + `catch (std::exception&)` fallback (after the existing `InternalError` arm), so non-categorized exceptions no longer get quietly absorbed as load errors.

2. `test/ebpf_yaml.cpp:375-380` — Narrowed the expected-exception matcher to `catch (RuntimeInputError&)` + `catch (YAML::Exception&)` so `InternalError` escapes upward instead of satisfying an `expected-exception` clause. The schema-validation throws inside `parse_observations`/`raw_options_to_options` were retagged from bare `std::runtime_error` to `RuntimeInputError` to keep the narrowed catch matching them.

3. `string_constraints.hpp` outside-diff — `contains()` now routes through the `value()` accessor so the bottom check goes through `CRAB_ERROR` consistently with the other bottom-state accesses, instead of falling through to `std::bad_optional_access`.

Tests still pass at baseline (1320 passed + 236 failed-as-expected + 1 skipped).
